src/main/scala/com/fulcrumgenomics/bam/TrimPrimers.scala

```diff
@@ -85,71 +91,117 @@ class TrimPrimers
+        val progress = new ProgressLogger(this.logger, "Written")
 
-      reads.foreach(out.addAlignment)
+        recs.foreach { rec =>
+          recalculateTags(rec, walker)
+          out.addAlignment(rec)
+          progress.record(rec)
+        }
+      case _ => Unit
     }
 
     out.close()
   }
 
+  /** Recalculates the NM, UQ and MD tags on aligned records. */
+  def recalculateTags(rec: SAMRecord, walker: ReferenceSequenceFileWalker): Unit = {
+    if (!rec.getReadUnmappedFlag) {
+      val refBases = walker.get(rec.getReferenceIndex).getBases
+      SequenceUtil.calculateMdAndNmTags(rec, refBases, true, true)
+      if (rec.getBaseQualities != null && rec.getBaseQualities.length != 0) {
+        rec.setAttribute(SAMTag.UQ.name, SequenceUtil.sumQualitiesOfMismatches(rec, refBases, 0))
+      }
+    }
+  }
+
+  /** Trims all the reads for a given template. */
+  def trimReadsForTemplate(detector: OverlapDetector[Amplicon], maxPrimerLength: Int, reads: Seq[SAMRecord]): Unit = {
+    val rec1 = reads.find(r => r.getReadPairedFlag && r.getFirstOfPairFlag && !r.isSecondaryOrSupplementary)
+    val rec2 = reads.find(r => r.getReadPairedFlag && r.getSecondOfPairFlag && !r.isSecondaryOrSupplementary)
+
+    (rec1, rec2) match {
+      case (Some(r1), Some(r2)) =>
+        // FR mapped pairs get the full treatment
+        if (!r1.getReadUnmappedFlag && !r2.getReadUnmappedFlag &&
+          r1.getReferenceName == r2.getReferenceName &&
+          SamPairUtil.getPairOrientation(r1) == PairOrientation.FR) {
+
+          val (left, right) = if (r1.getReadNegativeStrandFlag) (r2, r1) else (r1, r2)
+          val (start, end) = (left.getUnclippedStart, right.getUnclippedEnd)
+          val insert = new Interval(left.getContig, start, end)
+          detector.getOverlaps(insert).find(amp => abs(amp.leftStart - start) <= slop && abs(amp.rightEnd - end) <= slop) match {
+            case Some(amplicon) =>
+              val leftClip = amplicon.leftPrimerLength
+              val rightClip = amplicon.rightPrimerLength
+              reads.foreach { rec =>
+                val toClip = if (rec.getFirstOfPairFlag == left.getFirstOfPairFlag) leftClip else rightClip
+                SamRecordClipper.clip5PrimeEndOfRead(rec, toClip, mode)
+              }
+            case None =>
+              reads.foreach(r => SamRecordClipper.clip5PrimeEndOfRead(r, maxPrimerLength, mode))
+          }
+
+          clipFullyOverlappedFrReads(r1, r2)
+        }
+        // Pairs without both reads mapped in FR orientation are just maximally clipped
+        else {
+          reads.foreach(r => SamRecordClipper.clip5PrimeEndOfRead(r, maxPrimerLength, mode))
+        }
+
+        SamPairUtil.setMateInfo(r1, r2, true)
+        reads.filter(_.getSupplementaryAlignmentFlag).foreach { rec =>
+          val mate = if (rec.getFirstOfPairFlag) r2 else r1
+          SamPairUtil.setMateInformationOnSupplementalAlignment(rec, mate, true);
+        }
```

Comment:
I was going to suggest using `SetMateInfoIterator` but I don't think it handles frags.